src/github.com/getlantern/flashlight/ui/service.go

```diff
@@ -9,11 +10,22 @@ type envelopeType struct {
 	Type string
 }
 
+// envelope is a struct that wraps messages and associates them with a type.
+type envelope struct {
+	envelopeType
+	Message interface{}
+}
+
+type helloFnType func(func(interface{}) error) error
+
 type Service struct {
-	Name    string
-	In      chan []byte
-	Out     chan []byte
-	helloFn func(func([]byte) error) error
+	Type       string
+	In         <-chan interface{}
+	Out        chan<- interface{}
+	in         chan interface{}
+	out        chan interface{}
+	newMessage func() interface{}
+	helloFn    helloFnType
```

Comment:
Several changes here.  Rather than passing around bytes, we now pass around marshalled data structures.  That way, the services themselves don't have to deal with the envelopes.

Also, I made private and public channels.  The public channels are one-way.  This is a nice pattern to prevent clients from writing to a channel that's only meant for them to read, or vice versa.